src/Lean/Meta/Tactic/Simp/Main.lean

```diff
@@ -723,7 +723,41 @@ where
     if e.isArrow then
       simpArrow e
     else if (← isProp e) then
-      withLocalDecl e.bindingName! e.bindingInfo! e.bindingDomain! fun x => withNewLemmas #[x] do
+      /- The forall is a proposition. -/
+      let domain := e.bindingDomain!
+      if (← isProp domain) then
+        /-
+        The domain of the forall is also a proposition, and we can use `forall_prop_domain_congr`
+        IF we can simplify the domain.
+        -/
+        let rd ← simp domain
+        if let some h₁ := rd.proof? then
+          /- Using
+          ```
+          theorem forall_prop_domain_congr {p₁ p₂ : Prop} {q₁ : p₁ → Prop} {q₂ : p₂ → Prop}
+              (h₁ : p₁ = p₂)
+              (h₂ : ∀ a : p₂, q₁ (h₁.substr a) = q₂ a)
+              : (∀ a : p₁, q₁ a) = (∀ a : p₂, q₂ a)
+          ```
+          Remark: we should consider whether we want to add congruence lemma support for arbitrary `forall`-expressions.
+          Then, the theroem above can be marked as `@[congr]` and the following code deleted.
+          -/
```

Comment:
Probably you already saw https://github.com/leanprover/lean4/issues/1926#issuecomment-1759397589, but in case you didn't; I stumbled across what might be a neat trick to enable this behavior (and probably also arrow support in `calc`); define
```lean
-- an alias for `∀` to thwart checks that complain that `∀` is not a function
def Forall {α} (β : α → Sort _) := ∀ a, β a
```
and temporarily replace `∀`  with `Forall` when looking for congruence lemmas / `Trans` instances, such that the existing code-path for `Expr.app` can be reused.